profiling/service/service.go

```diff
@@ -0,0 +1,123 @@
+/*
+ *
+ * Copyright 2019 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// The service package defines methods to register a gRPC client/service for a
+// profiling service that is exposed in the same server. This service can be
+// queried by a client to remotely manage the gRPC profiling behaviour of an
+// application.
+package service
+
+import (
+	"context"
+	"errors"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/grpclog"
+
+	"google.golang.org/grpc/internal/profiling"
+	pspb "google.golang.org/grpc/profiling/proto/service"
+)
+
+// Profiling configuration options.
+type ProfilingConfig struct {
+	// Setting this to true will enable profiling.
+	Enabled bool
+
+	// Profiling uses a circular buffer (ring buffer) to store statistics for
+	// only the last few RPCs so that profiling stats do not grow unbounded. This
+	// parameter defines the upper limit on the number of RPCs for which
+	// statistics should be stored at any given time. An average RPC requires
+	// approximately 2-3 KiB of memory for profiling-related statistics, so
+	// choose an appropriate number based on the amount of memory you can afford.
+	StreamStatsSize uint32
+
+	// To expose the profiling service and its methods, a *grpc.Server must be
+	// provided.
+	Server *grpc.Server
+}
+
+var errorNilServer = errors.New("No grpc.Server provided.")
+
+// Init takes a *ProfilingConfig to initialise profiling (turned on/off
+// depending on the value set in pc.Enabled) and register the profiling service
+// in the server provided in pc.Server.
+func Init(pc *ProfilingConfig) error {
```

Comment:
Does `ProfilingConfig` need to be a pointer here?

Reply:
Passing a non-pointer type struct to a function requires copying the whole thing, no? While the struct is certainly small right now and shouldn't be any issue, IMO we should avoid doing that for the future's sake.

Also, I noticed a similar pattern in the gRPC code (especially in the transport package; for e.g., a `*transport.Options` is passed to `http2Server.Write` and `http2Client.Write`).

I don't really mind doing either FWIW, so let me know if you still want me to change this.

Reply:
The readability doc has the following section:
```
Don't pass pointers as function arguments just to save a few bytes. If a function refers to its argument x only as *x throughout, then the argument shouldn't be a pointer. Common instances of this include passing a pointer to a string (*string) or a pointer to an interface value (*io.Reader). In both cases the value itself is a fixed size and can be passed directly. This advice does not apply to large structs, or even small structs that might grow.
```
I'm ok leaving it as is for now, but based on my other comments, things might change :)